src/lib.rs

```diff
@@ -398,13 +409,123 @@ impl Builder {
+        );
+        let out_dir = self.out_dir;
+
+        let mut include_paths = self.include_paths;
+        for path in &mut include_paths {
+            println!("cargo:rerun-if-changed={}", path.display());
+            let destination =
+                out_dir.join(path.file_name().expect("include path to have filename"));
+            std::fs::copy(path.clone(), destination).expect("copy include headers");
+            // remove the filename from the path so it's just the directory
+            path.pop();
+        }
+
+        include_paths.sort();
+        include_paths.dedup();
+
+        #[allow(unused)]
+        let mut include_options: Vec<String> = include_paths
+            .into_iter()
+            .map(|s| {
+                "-I".to_string()
+                    + &s.into_os_string()
+                        .into_string()
+                        .expect("include option to be valid string")
+            })
+            .collect::<Vec<_>>();
+        include_options.push(format!("-I{}", cuda_include_dir.display()));
+
+        let ccbin_env = std::env::var("NVCC_CCBIN");
+        println!("cargo:rerun-if-env-changed=NVCC_CCBIN");
+        for path in &self.watch {
+            println!("cargo:rerun-if-changed={}", path.display());
+        }
+        let children = self.kernel_paths
+            .par_iter()
+            .flat_map(|p| {
+                println!("cargo:rerun-if-changed={}", p.display());
+                let mut output = p.clone();
+                output.set_extension("cubin");
+                let output_filename = std::path::Path::new(&out_dir).to_path_buf().join("out").with_file_name(output.file_name().expect("kernel to have a filename"));
+
+                let ignore = if let Ok(metadata) = output_filename.metadata() {
+                    let out_modified = metadata.modified().expect("modified to be accessible");
+                    let in_modified = p.metadata().expect("input to have metadata").modified().expect("input metadata to be accessible");
+                    out_modified.duration_since(in_modified).is_ok()
+                } else {
+                    false
+                };
+                if ignore {
+                    None
+                } else {
+                    let mut command = std::process::Command::new("nvcc");
+                    command.arg(format!("--gpu-architecture=sm_{compute_cap}"))
+                        .arg("--cubin")
+                        .args(["--default-stream", "per-thread"])
+                        .args(["--output-directory", &out_dir.display().to_string()])
+                        .args(&self.extra_args)
+                        .args(&include_options);
+                    if let Ok(ccbin_path) = &ccbin_env {
+                        command
+                            .arg("-allow-unsupported-compiler")
+                            .args(["-ccbin", ccbin_path]);
+                    }
+                    command.arg(p);
```

Comment:
Rather than `.args(&include_options)`, we could correct a possible crash on non-UTF-8 paths (avoids `.into_string()`) by directly passing the paths to the command via `AsRef<OsStr>` (idem for `build_ptx`).
```
        for include in &include_paths {
            command.arg("-I").arg(include);
        }
        command.arg("-I").arg(&cuda_include_dir);
        command.arg(p);
```